src/v/datalake/arrow_schema_translator.cc

```diff
@@ -0,0 +1,273 @@
+    arrow::FieldVector _fields;
+    std::shared_ptr<arrow::DataType> _arrow_data_type;
+};
+
+class list_schema_converter : public schema_converter_interface {
+public:
+    explicit list_schema_converter(const iceberg::list_type& schema);
+
+    std::shared_ptr<arrow::Field>
+    field(const std::string& name, iceberg::field_required required) override {
+        return arrow::field(
+          name, _arrow_data_type, required == iceberg::field_required::no);
+    }
+
+private:
+    std::shared_ptr<arrow::DataType> _arrow_data_type;
+};
+
+class map_schema_converter : public schema_converter_interface {
+public:
+    explicit map_schema_converter(const iceberg::map_type& schema);
+
+    std::shared_ptr<arrow::Field>
+    field(const std::string& name, iceberg::field_required required) override {
+        if (!_arrow_data_type) {
+            return nullptr;
+        }
+        return arrow::field(
+          name, _arrow_data_type, required == iceberg::field_required::no);
+    }
+
+private:
+    std::shared_ptr<arrow::DataType> _arrow_data_type;
+};
+
+struct schema_converter_builder_visitor {
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::struct_type& s) {
+        return std::make_unique<struct_schema_converter>(s);
+    }
+
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::list_type& s) {
+        return std::make_unique<list_schema_converter>(s);
+    }
+
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::map_type& m) {
+        return std::make_unique<map_schema_converter>(m);
+    }
+
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::primitive_type& p) {
+        return std::visit(*this, p);
+    }
+
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::boolean_type& /*t*/) {
+        return std::make_unique<scalar_schema_converter<arrow::BooleanType>>(
+          arrow::boolean());
+    }
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::int_type& /*t*/) {
+        return std::make_unique<scalar_schema_converter<arrow::Int32Type>>(
+          arrow::int32());
+    }
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::long_type& /*t*/) {
+        return std::make_unique<scalar_schema_converter<arrow::Int64Type>>(
+          arrow::int64());
+    }
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::float_type& /*t*/) {
+        return std::make_unique<scalar_schema_converter<arrow::FloatType>>(
+          arrow::float32());
+    }
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::double_type& /*t*/) {
+        return std::make_unique<scalar_schema_converter<arrow::DoubleType>>(
+          arrow::float64());
+    }
+    std::unique_ptr<schema_converter_interface>
+    operator()(const iceberg::decimal_type& t) {
+        if (t.precision == 0) {
+            return nullptr;
```

Comment:
nit: would it make sense to throw instead? Is the expectation that callers will need to check for nullptr?